src/block_data.rs

```diff
@@ -0,0 +1,320 @@
+}
+
+
+/// A data structure that can be converted from [hotstuff_rs::types::Data].
+pub struct BlockDataV2 {
+    pub header: BlockHeaderDataV2,
+    pub transactions: Vec<TransactionV2>,
+    pub receipts: Vec<ReceiptV2>,
+}
+
+impl BlockDataV2 {
+
+    /// Conversion from [hotstuff_rs::types::Data] with an option to validate the transactions.
+    pub fn from_data(
+        hotstuff_data: &Data,
+        verify_transaction_signatures: bool
+    ) -> Result<BlockDataV2, BlockConversionError> {
+        // Construct BlockDataheader from fixed datum indexes 
+        let header: BlockHeaderDataV2 = BlockHeaderDataV2::try_from(hotstuff_data)
+            .map_err(BlockConversionError::WrongHeader)?;
+        
+        // Construct Transactions and Receipts from dynamic datum indexes
+        let (txns_bs, receipts_bs) = DatumIndexV2::transactions_and_receipts(hotstuff_data)
+            .ok_or(BlockConversionError::IncorrectNumberOfTxsAndReceipts)?;
+
+        // Deserialize transactions (and validate the signatures)
+        let mut transactions = Vec::with_capacity(txns_bs.len());
+        for txn_bs in txns_bs {
+            let txn: TransactionV2 = Deserializable::deserialize(txn_bs)
+                .map_err(|_| BlockConversionError::Transaction)?;
+            // Check transactions signature and return error immediately if there is an invalid transaction.
+            if verify_transaction_signatures {
+                txn.is_cryptographically_correct()
+                    .map_err(|_| BlockConversionError::InvalidTransactionSignature)?;
+            }
+            transactions.push(txn);
+        }
+
+        // Deserialize receipts
+        let mut receipts: Vec<ReceiptV2> = Vec::with_capacity(receipts_bs.len());
+        for receipt_bs in receipts_bs {
+            receipts.push(
+                Deserializable::deserialize(receipt_bs)
+                    .map_err(|_| BlockConversionError::Receipt)?,
+            )
+        }
+
+        Ok(BlockDataV2{
+            header,
+            transactions,
+            receipts,
+        })
+    }
+}
+
+impl From<BlockDataV2> for Data {
+    fn from(value: BlockDataV2) -> Data {
+        let header = Data::from(value.header);
+
+        let transactions: Data = value
+            .transactions
+            .into_iter()
+            .map(|tx| Serializable::serialize(&tx))
+            .collect();
+
+        let receipts: Data = value
+            .receipts
+            .into_iter()
+            .map(|recp| Serializable::serialize(&recp))
+            .collect();
+
+        [header, transactions, receipts].concat()
+    }
+}
+
+impl<'a> TryFrom<&'a Data> for BlockDataV2 {
+    type Error = BlockConversionError;
+    fn try_from(data: &'a Data) -> std::result::Result<Self, Self::Error> {
+        BlockDataV2::from_data(data, true)
+    }
+}
+
+
+/// A data structure that encapsulates the data of block header from [hotstuff_rs::types::Data].
+#[derive(Clone, BorshSerialize, BorshDeserialize)]
```

Comment:
I wonder why we need to BlockHeaderDataV2 to implement BorshSerialize, BorshDeserialize, and Serializable.

Reply:
The traits are removed now because those are not useful (so far).